### PR TITLE
Make the view focus binding report focus transitions across elements.

### PR DESCRIPTION
**Pull request by tugorez**
Make the view focus binding report focus transitions across elements.

Previously the web engine reported all the focusin/focusout events as if the elements were first blurred before moving the focus.

Relevant Issues are:

* Design doc: https://flutter.dev/go/focus-management 
* Focus in web multiview: https://github.com/flutter/flutter/issues/137443
* Platform dispatcher changes: https://github.com/flutter/engine/pull/49841

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart [hunk 1]
```diff
@@ -39,9 +39,17 @@ final class ViewFocusBinding {
     }
   }
 
+  late final DomEventListener _handleFocusin = createDomEventListener(
+    (DomEvent event) => _handleFocusChange(event.target as DomElement?),
+  );
+
+  late final DomEventListener _handleFocusout = createDomEventListener(
+    (DomEvent event) => _handleFocusChange(event.relatedTarget as DomElement?),
+  );
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:48]: Reminds me of [something](https://github.com/flutter/engine/pull/50177#discussion_r1480875921)... :P

File: lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart [hunk 2]
```diff
@@ -86,19 +86,15 @@ void testMain() {
...
+      expect(viewFocusEvents, hasLength(2));
 
       expect(viewFocusEvents[0].viewId, view1.viewId);
       expect(viewFocusEvents[0].state, ui.ViewFocusState.focused);
       expect(viewFocusEvents[0].direction, ui.ViewFocusDirection.forward);
 
-      expect(viewFocusEvents[1].viewId, view1.viewId);
-      expect(viewFocusEvents[1].state, ui.ViewFocusState.unfocused);
-      expect(viewFocusEvents[1].direction, ui.ViewFocusDirection.undefined);
-
```
Comment (ditman) [inline lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart:98]: How come we've lost one event with this change? The listeners are almost the same?

I would expect to have:

* view1 focused
* view2 focused
* view1 unfocused

?

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart (hunk 1 above)
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:48]: Great catch. No wonder you're a true pro!

Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:48]: Will elab in the comment below on why this change is needed.

File: lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart (hunk 2 above)
Comment (tugorez) [reply lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart:98]: Hmmmm maybe I'm missing something; but don't think that's what the browser do. It first calls out the blur events and then the focus one so it would be more like the behavior that's being changed:

1. view 1 focus
2. view 1 unfocus
3. view 2 focus

This is problematic due to the way flutter handles the semantic nodes & text fields. As soon as the user moves out of 
 a text field or semantic node (no matter what's next) it blurs the node and `2)` happens. When `2)` happens it tells the framework to unfocus everything then `3)` happens. When `3` happens, right after, we tell the fraemwork to focus the view back and pick the first focusable widget (node). With this change, even if the input is blurred we let the framework handle the focus traversal by not telling it to unnecesarily focus/unfocus the view.

Comment (ditman) [reply lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart:98]: Let me rephrase:

Why did we have 3 events before, and now only 2, if we're using the same event listeners? :P

Comment (tugorez) [reply lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart:98]: activeElement is either body or null as it transitions across views hence it was interpreted as you were blurring before focusing the view

Comment (tugorez) [reply lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart:98]: We were previously using `activeElement`. In this change we use the event `target` and `relatedTarget` respectively.

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart [hunk 3]
```diff
@@ -39,9 +39,17 @@ final class ViewFocusBinding {
     }
   }
 
+  late final DomEventListener _handleFocusin = createDomEventListener(
+    (DomEvent event) => _handleFocusChange(event.target as DomElement?),
+  );
+
+  late final DomEventListener _handleFocusout = createDomEventListener(
+    (DomEvent event) => _handleFocusChange(event.relatedTarget as DomElement?),
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:47]: `relatedTarget` should be added to a `DomFocusEvent` class, it doesn't exist in the root `DomEvent` class. See:

* https://developer.mozilla.org/en-US/docs/Web/API/Event/Comparison_of_Event_Targets
* https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent#instance_properties

**Review by ditman — APPROVED** [review]
LGTM! Thanks for keeping the dom.dart sane!

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart (hunk 3 above)
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart]: Done
